File: sidecar-app/backend/fileCleaner.js
const path = require("path");
const fs = require("fs-extra");

module.exports = function (app) {
  /***
   * getFilesInFolder - get files in folder use post method
   * @param {string} folder - folder to get files from
   * @returns {object} - response from server
   *
   * @example
   * getFilesInFolder("D:\\Pictures")
   */
  app.post("/getFilesInFolder", async (req, res) => {
    const folder = req.body.folder;

    try {
      fs.readdir(folder, { withFileTypes: true }, (err, files) => {
        if (err) {
          res.status(500).send(err.message);
        } else {
          const fileList = files
            .filter((file) => file.isFile())
            .map((file) => {
              return {
                filename: file.name,
                path: path.join(folder, file.name),
              };
            });
          res.status(200).send(fileList);
        }
      });
    } catch (err) {
      res.status(500).send(err.message);
    }
  });

  /*** rename files in folder use post method renameFilesInFolder
   * @param FileCleanerProps [{ path: string; filename: string; fixed: string;},...]
   * @returns FileCleanerProps [{ path: string; filename: string; fixed:string, status: string;},...]
   * @example renameFilesInFolder([{ path: "D:\\Pictures\\IMG_0001.jpg", filename: "IMG_0001.jpg", fixed: "IMG_0001.jpg" },...])
   * */
  app.post("/renameFilesInFolder", async (req, res) => {
    const files = req.body;
    const results = [];

    try {
      const renamePromises = files.map(async (file) => {
        try {
          const newPath = path.join(path.dirname(file.path), file.fixed);
          await fs.rename(file.path, newPath);
          results.push({ ...file, status: "success" });
        } catch (err) {
          results.push({ ...file, status: "error", message: err.message });
        }
      });
      await Promise.all(renamePromises);
      res.status(200).send(results);
    } catch (err) {
      res.status(500).send(err.message);
    }
  });
};

File: src-tauri/src/file_cleaner.rs
use tauri::command;
use std::{fs, path::Path};

#[derive(serde::Deserialize, serde::Serialize)]
pub struct FileCleanerProps {
    pub path: String,
    pub filename: String,
    pub fixed: Option<String>,
    pub status: Option<String>,
    pub message: Option<String>,
}

#[command]
pub fn get_files_in_folder(folder: String) -> Result<Vec<FileCleanerProps>, String> {
    let mut results = Vec::new();
    let entries = fs::read_dir(&folder).map_err(|e| e.to_string())?;

    for entry in entries {
        let entry = entry.map_err(|e| e.to_string())?;
        let path = entry.path();
        if path.is_file() {
            let filename = path
                .file_name()
                .unwrap_or_default()
                .to_string_lossy()
                .to_string();
            results.push(FileCleanerProps {
                path: path.to_string_lossy().to_string(),
                filename,
                fixed: None,
                status: None,
                message: None,
            });
        }
    }
    Ok(results)
}

#[command]
pub fn rename_files_in_folder(files: Vec<FileCleanerProps>) -> Result<Vec<FileCleanerProps>, String> {
    let mut results = Vec::new();

    for mut file_info in files {
        if let Some(ref new_name) = file_info.fixed {
            let parent = Path::new(&file_info.path).parent();
            if let Some(parent_path) = parent {
                let new_path = parent_path.join(new_name);
                match fs::rename(&file_info.path, &new_path) {
                    Ok(_) => {
                        file_info.path = new_path.to_string_lossy().to_string();
                        file_info.status = Some("success".to_string());
                    }
                    Err(e) => {
                        file_info.status = Some("error".to_string());
                        file_info.message = Some(e.to_string());
                    }
                }
            } else {
                file_info.status = Some("error".to_string());
                file_info.message = Some("Invalid path parent".to_string());
            }
        }
        results.push(file_info);
    }
    Ok(results)
}
